server/message_hooks_test.go

```diff
@@ -822,7 +821,299 @@ func TestReactionHasBeenRemoved(t *testing.T) {
 	})
 }
 
+func TestMessageHasBeenPosted(t *testing.T) {
+	th := setupTestHelper(t)
+	team := th.SetupTeam(t)
+
+	expectChat := func(th *testHelper, t *testing.T, users ...*model.User) {
+		t.Helper()
+
+		var members []clientmodels.ChatMember
+		for _, user := range users {
+			members = append(members, clientmodels.ChatMember{
+				UserID:      "t" + user.Id,
+				Email:       user.Email,
+				DisplayName: user.GetDisplayName(""),
+			})
+		}
+
+		chatID := model.NewId()
+		th.clientMock.On("CreateOrGetChatForUsers", mock.AnythingOfType("[]string")).Return(&clientmodels.Chat{
+			ID:      chatID,
+			Members: members,
+		}, nil).Maybe()
+
+		teamsMessageID := model.NewId()
+		th.clientMock.On(
+			"SendChat",
+			chatID,
+			mock.AnythingOfType("string"),
+			(*clientmodels.Message)(nil),
+			[]*clientmodels.Attachment(nil),
+			[]models.ChatMessageMentionable{},
+		).Return(&clientmodels.Message{
+			ID:       teamsMessageID,
+			CreateAt: time.Now(),
+		}, nil).Times(1)
+	}
+
+	expectChatSync := func(t *testing.T) {
+		t.Helper()
+
+		assert.Eventually(t, func() bool {
+			return th.getRelativeCounter(t,
+				"msteams_connect_events_messages_total",
+				withLabel("action", metrics.ActionCreated),
+				withLabel("source", metrics.ActionSourceMattermost),
+				withLabel("is_direct", "true"),
+			) == 1
+		}, 1*time.Second, 250*time.Millisecond)
+	}
+
+	expectNoChatSync := func(t *testing.T) {
+		t.Helper()
+
+		assert.Never(t, func() bool {
+			return th.getRelativeCounter(t,
+				"msteams_connect_events_messages_total",
+				withLabel("action", metrics.ActionCreated),
+				withLabel("source", metrics.ActionSourceMattermost),
+				withLabel("is_direct", "true"),
+			) == 1
+		}, 1*time.Second, 250*time.Millisecond)
+	}
+
+	t.Run("MS Teams bot ignored", func(t *testing.T) {
+		t.Skip("Not yet implemented")
+	})
+
+	t.Run("system message ignored", func(t *testing.T) {
+		t.Skip("Not yet implemented")
+	})
+
+	type parameters struct {
+		SenderConnected    bool
+		SelectiveSync      bool
+		SyncDirectMessages bool
+		SyncGroupMessages  bool
+		SyncLinkedChannels bool
+	}
+	runPermutations(t, parameters{}, func(t *testing.T, params parameters) {
+		th.Reset(t)
+
+		th.setPluginConfiguration(t, func(c *configuration) {
+			c.SelectiveSync = params.SelectiveSync
+			c.SyncDirectMessages = params.SyncDirectMessages
+			c.SyncGroupMessages = params.SyncGroupMessages
+			c.SyncLinkedChannels = params.SyncLinkedChannels
+		})
+
+		t.Run("dm with self", func(t *testing.T) {
+			th.Reset(t)
+
+			user1 := th.SetupUser(t, team)
+
+			client1 := th.SetupClient(t, user1.Id)
+
+			channel, _, err := client1.CreateDirectChannel(context.TODO(), user1.Id, user1.Id)
+			require.NoError(t, err)
+
+			if params.SenderConnected {
+				th.ConnectUser(t, user1.Id)
+			}
```

Comment:
And when testing, I can just use the permutation parameters to properly setup the test.